python/ray/data/_internal/planner/plan_expression/expression_visitors.py

```diff
@@ -0,0 +1,265 @@
+from typing import Set
+
+from ray.data.expressions import (
+    AliasExpr,
+    BinaryExpr,
+    ColumnExpr,
+    Expr,
+    LiteralExpr,
+    StarExpr,
+    UDFExpr,
+    UnaryExpr,
+    _ExprVisitor,
+)
+
+
+class _ColumnReferenceCollector(_ExprVisitor[None]):
+    """Visitor that collects all column references from expression trees.
+
+    This visitor traverses expression trees and accumulates column names
+    referenced in ColumnExpr nodes.
+    """
+
+    def __init__(self):
+        """Initialize with an empty set of referenced columns."""
+        self.referenced_columns: Set[str] = set()
+
+    def visit(self, expr: Expr) -> None:
+        """Visit an expression node and dispatch to the appropriate method.
+
+        Extends the base visitor to handle StarExpr which is not
+        part of the base _ExprVisitor interface.
+
+        Args:
+            expr: The expression to visit.
+
+        Returns:
+            None (only collects columns as a side effect).
+        """
+        super().visit(expr)
+
+    def visit_column(self, expr: ColumnExpr) -> None:
+        """Visit a column expression and collect its name.
+
+        Args:
+            expr: The column expression.
+
+        Returns:
+            None (only collects columns as a side effect).
+        """
+        self.referenced_columns.add(expr.name)
+
+    def visit_literal(self, expr: LiteralExpr) -> None:
+        """Visit a literal expression (no columns to collect).
+
+        Args:
+            expr: The literal expression.
+
+        Returns:
+            None.
+        """
+        # Literals don't reference any columns
+        pass
+
+    def visit_binary(self, expr: BinaryExpr) -> None:
+        """Visit a binary expression and collect from both operands.
+
+        Args:
+            expr: The binary expression.
+
+        Returns:
+            None (only collects columns as a side effect).
+        """
+        self.visit(expr.left)
+        self.visit(expr.right)
```

Comment:
nit: Create `_ExprVisitorBase`  and make it traverse whole AST recursively to reduce boilerplate in every visitor